### PR TITLE
Ading Proxy template support in zanzibar

### PR DESCRIPTION
**Pull request by shivam-srivastava28**

### CODE REVIEW

File: codegen/gateway.go [hunk 1]
```diff
@@ -401,6 +403,12 @@ func NewEndpointSpec(
 			thriftInfo, yamlFile,
 		)
 	}
+	var config map[string]interface{}
+	if _, ok := endpointConfigObj["config"]; !ok {
+		config = make(map[string]interface{})
+	} else {
+		config = endpointConfigObj["config"].(map[string]interface{})
```
Comment (hbhoyar-uber) [inline codegen/gateway.go:410]: can you do 
`config, ok := ....`

File: codegen/module.go [hunk 2]
```diff
@@ -83,6 +83,7 @@ type Options struct {
 	EnableCustomInitialisation bool
 	CommitChange               bool
 	QPSLevelsEnabled           bool
+	ProxyTemplates             *Template
```
Comment (groshu) [inline codegen/module.go:86]: maybe a better name like custom templates to make it generic ? just a thought no hard opinions

File: codegen/template.go [hunk 3]
```diff
@@ -139,6 +142,24 @@ func NewTemplate(
...
+	defaultTemplate *Template,
+	proxyTemplate *Template,
+	config map[string]interface{},
+) (string, *Template) {
+	// verify if config contains template override
+	if _, ok := config[ProxyTemplate]; !ok {
+		return defaultTemplateName, defaultTemplate
+	}
+	// verify if override template config contains given template
+	templateMap := config[ProxyTemplate].(map[string]interface{})
```
Comment (hbhoyar-uber) [inline codegen/template.go:156]: can you do 
`templateMap, ok :=  ...`

File: codegen/module.go [hunk 4]
```diff
@@ -1787,6 +1789,8 @@ type ModuleInstance struct {
 	mu                sync.RWMutex
 	// QPSLevels is map of circuit breaker name to qps level for all circuit breakers
 	QPSLevels map[string]int
+	// ProxyTemplates is a collection of proxy templates apart from zanzibar
```
Comment (groshu) [inline codegen/module.go:1792]: if using custom templates. we can write collection of custom templates that can be passed to zanzibar

**Review by hbhoyar-uber — CHANGES_REQUESTED** [review]
Can you also add this new config to one of the examples to verify the change

**Review by groshu — COMMENTED** [review]
can we please add unit tests ?

File: codegen/module.go (hunk 4 above)
Comment (shivam-srivastava28) [reply codegen/module.go]: Done

File: codegen/module.go (hunk 2 above)
Comment (shivam-srivastava28) [reply codegen/module.go]: Done

File: codegen/template.go (hunk 3 above)
Comment (shivam-srivastava28) [reply codegen/template.go]: Same as above

**Comment by shivam-srivastava28** [comment]
> 

Added unit test cases

**Comment by groshu** [comment]
should we assert a few things like in the test template default template is not picked but passed custom template is picked ?
is it easy to do that ?

**Review by groshu — APPROVED** [review]

**Review by hbhoyar-uber — APPROVED** [review]

**Pull request merged**